tools/c7n_gcp/c7n_gcp/actions/iampolicy.py

```diff
@@ -0,0 +1,224 @@
+        are empty, the possibility of which is defined by the caller of the method.
+
+        For additional information on how the method works, please refer to the tests
+        (e.g. test_spanner).
+
+        :param existing_bindings: a list of dictionaries containing the 'role' and 'members' keys
+                                  taken from the resource the action is applied to
+        :param bindings_to_add: a list of dictionaries containing the 'role' and 'members' keys
+                                taken from the policy
+        """
+        bindings = []
+        roles_to_existing_bindings = self._get_roles_to_bindings_dict(existing_bindings)
+        roles_to_bindings_to_add = self._get_roles_to_bindings_dict(bindings_to_add)
+        for role in roles_to_bindings_to_add:
+            updated_members = dict(roles_to_bindings_to_add[role])
+            if role in roles_to_existing_bindings:
+                existing_members = roles_to_existing_bindings[role]['members']
+                members_to_add = list(filter(lambda member: member not in existing_members,
+                                             updated_members['members']))
+                updated_members['members'] = existing_members + members_to_add
+            bindings.append(updated_members)
+
+        for role in roles_to_existing_bindings:
+            if role not in roles_to_bindings_to_add:
+                bindings.append(roles_to_existing_bindings[role])
+        return bindings
+
+    def _remove_bindings(self, existing_bindings, bindings_to_remove):
+        """
+        Converts the provided lists using `_get_roles_to_bindings_dict`, then iterates through
+        them so that the returned list combines:
+        - among the roles mentioned in a policy, only the members that are not marked for removal,
+        - as for the other roles, all their members.
+
+        The roles or members that are mentioned in the policy but are absent
+        in the existing bindings are simply ignored with no errors produced.
+
+        As can be observed, it is possible to have an empty list returned either if
+        `existing_bindings` is already empty or `bindings_to_remove` filters everything out.
+
+        For additional information on how the method works, please refer to the tests
+        (e.g. test_spanner).
+
+        :param existing_bindings: a list of dictionaries containing the 'role' and 'members' keys
+                                  taken from the resource the action is applied to
+        :param bindings_to_remove: a list of dictionaries containing the 'role' and 'members' keys
+                                   taken from the policy
+        """
+        bindings = []
+        roles_to_existing_bindings = self._get_roles_to_bindings_dict(existing_bindings)
+        roles_to_bindings_to_remove = self._get_roles_to_bindings_dict(bindings_to_remove)
+        for role in roles_to_bindings_to_remove:
+            if role in roles_to_existing_bindings:
```

Comment:
this syntax doesn't really allow for partial, its only exact match, ie if you want to remove admin role or read-only role you have to know the members, ideally you could just set the role to remove perhaps with a '*' for members

also the value being iterated is not a role its a binding dictionary with a role attribute.

Reply:
Thank you for the advice, now `members: *` is available to remove all role members.
Speaking about the code, in this case keys are iterated thus `role` - the key in a dictionary - looks appropriate enough.

Reply:
thanks, looks pretty good, one minor on the doc example noted later.